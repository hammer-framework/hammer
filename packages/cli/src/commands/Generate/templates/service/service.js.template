import { Photon } from '@prisma/photon'

const photon = new Photon()

const ${pluralName} = {
  ${pluralCamelName}: () => {
    return photon.${pluralCamelName}.findMany()
  },
<% if (isCrud) { %>
  ${singularCamelName}: ({ id }) => {
    return photon.${pluralCamelName}.findOne({
      where: { id: id },
    })
  },

  create${singularName}: ({ input }) => {
    return photon.${pluralCamelName}.create({
      data: input,
    })
  },

  update${singularName}: ({ id, input }) => {
    return photon.${pluralCamelName}.update({
      data: input,
      where: { id: id },
    })
  },

  delete${singularName}: ({ id }) => {
    return photon.${pluralCamelName}.delete({
      where: { id: id },
    })
  },<% } %>
}

export default ${pluralName}
