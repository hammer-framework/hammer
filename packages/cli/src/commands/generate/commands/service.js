import camelcase from 'camelcase'
import pluralize from 'pluralize'

import {
  templateForComponentFile,
  createYargsForComponentGeneration,
} from '../helpers'

export const files = async ({ name, ...rest }) => {
  const componentName = camelcase(pluralize(name))
  const serviceFile = templateForComponentFile({
    name,
    componentName: componentName,
    apiPathSection: 'services',
    templatePath: 'service/service.js.template',
    templateVars: { ...rest },
  })
  const testFile = templateForComponentFile({
    name,
    componentName: componentName,
    extension: '.test.js',
    apiPathSection: 'services',
    templatePath: 'service/test.js.template',
    templateVars: { ...rest },
  })

  // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }
  return [serviceFile, testFile].reduce((acc, [outputPath, content]) => {
    return {
      [outputPath]: content,
      ...acc,
    }
  }, {})
}

export const builder = {
  crud: { type: 'boolean', default: false, desc: 'Create CRUD functions' },
  force: { type: 'boolean', default: false },
}

export const { command, desc, handler } = createYargsForComponentGeneration({
  componentName: 'service',
  filesFn: files,
})
