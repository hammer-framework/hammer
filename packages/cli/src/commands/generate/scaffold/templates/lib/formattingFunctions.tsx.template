import React from 'react'

import humanize from 'humanize-string'

const MAX_STRING_LENGTH = 150

export const formatEnum = (values: string | string[] | null | undefined) => {
  if (values) {
    if (Array.isArray(values)) {
      const humanizedValues = values.map((value) => humanize(value))
      return humanizedValues.join(', ')
    } else {
      return humanize(values as string)
    }
  }

  return ''
}

export const truncate = (value: string | number) => {
  const output = value?.toString()
  if (output?.length > MAX_STRING_LENGTH) {
    return output.substring(0, MAX_STRING_LENGTH) + '...'
  }
  return output ?? ''
}

export const jsonTruncate = (obj: unknown) => {
  return truncate(JSON.stringify(obj, null, 2))
}

export const timeTag = (datetime?: string) => {
  return datetime ? (
    <time dateTime={datetime} title={datetime}>
      {new Date(datetime).toUTCString()}
    </time>
  ) : (
    ''
  )
}

export const checkboxInputTag = (checked: boolean) => {
  return <input type="checkbox" checked={checked} disabled />
}
