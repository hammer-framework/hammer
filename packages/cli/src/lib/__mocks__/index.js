export const getPaths = () => MOCK_PATHS

export const MOCK_PATHS = {
  base: 'mock_basedir',
  generated: {
    base: '.mock_redwood',
    schema: '.mock_redwood/schema.graphql',
    types: {
      includes: '.mock_redwood/types/includes',
      mirror: '.mock_redwood/types/mirror',
    },
    prebuild: '.mock_redwood/prebuild',
  },
  scripts: 'mock_scripts_dir',
  api: {
    base: 'mock_api',
    dataMigrations: 'mock_dataMigrations',
    db: 'mock_db_dir',
    dbSchema: 'mock_db_schema_dir',
    functions: 'fmock_unctions_dir',
    graphql: 'mock_graphql_dir',
    lib: 'mock_lib_dir',
    generators: 'mock_generators_dir',
    config: 'mock_config_dir',
    services: 'mock_services_dir',
    directives: 'mock_directives_dir',
    src: 'mock_src_dir',
    dist: 'mock_api/dist',
    types: 'mock_api/types',
    models: 'mock_models_dir',
  },
  web: {
    routes: ['mock_mock/routes'],
    base: 'mock_web',
    pages: 'mock_pages_dir',
    components: 'mock_components_dir',
    layouts: 'mock_layouts_dir',
    src: 'mock_src_dir',
    generators: 'mock_generators_dir',
    app: 'mock_app_dir',
    index: 'mock_index_dir',
    config: 'mock_config_dir',
    webpack: 'mock_webpack_dir',
    postcss: 'mock_postcss_dir',
    storybookConfig: 'mock_storybookConfig_dir',
    storybookPreviewConfig: 'mock_storybookPreviewConfig_dir',
    storybookManagerConfig: 'mock_storybookManagerConfig_dir',
    dist: 'mock_web/dist',
    types: 'mock_web/types',
  },
}
