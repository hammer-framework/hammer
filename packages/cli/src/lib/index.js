import fs from 'fs'
import path from 'path'

import lodash from 'lodash/string'
import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { paramCase } from 'param-case'
import { getDMMF } from '@prisma/sdk'
import { getPaths as getRedwoodPaths } from '@redwoodjs/internal'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import c from 'src/lib/colors'

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  if (name) {
    const model = metadata.datamodel.models.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw `No schema definition found for \`${name}\``
    }
  }

  return metadata.datamodel
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
*/
const nameVariants = (name) => {
  const normalizedName = pascalcase(pluralize.singular(name))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
  }
}

export const templateRoot = path.resolve(
  __dirname,
  '../commands/Generate/templates'
)

export const generateTemplate = (templateFilename, { name, ...rest }) => {
  const templatePath = path.join(templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())
  return template({
    name,
    ...nameVariants(name),
    ...rest,
  })
}

export const readFile = (target) => fs.readFileSync(target)

export const writeFile = async (
  target,
  contents,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(0)
  }
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `Writing \`./${path.relative(base, file)}\`...`,
        task: () => writeFile(file, contents, options),
      }
    })
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reduce((content, route) => {
    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
  }, routesContent)
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}

export * from './generateTempSchema'
