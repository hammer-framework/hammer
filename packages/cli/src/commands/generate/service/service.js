import camelcase from 'camelcase'
import pluralize from 'pluralize'

import { transformTSToJS } from '../../../lib'
import {
  templateForComponentFile,
  createYargsForComponentGeneration,
} from '../helpers'

export const files = async ({
  name,
  relations,
  javascript,
  typescript,
  ...rest
}) => {
  const componentName = camelcase(pluralize(name))
  const extension = 'ts'
  const serviceFile = templateForComponentFile({
    name,
    componentName: componentName,
    extension: `.${extension}`,
    apiPathSection: 'services',
    generator: 'service',
    templatePath: `service.${extension}.template`,
    templateVars: { relations: relations || [], ...rest },
  })
  const testFile = templateForComponentFile({
    name,
    componentName: componentName,
    extension: `.test.${extension}`,
    apiPathSection: 'services',
    generator: 'service',
    templatePath: `test.${extension}.template`,
    templateVars: { relations: relations || [], ...rest },
  })

  // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }
  return [serviceFile, testFile].reduce((acc, [outputPath, content]) => {
    if (javascript && !typescript) {
      content = transformTSToJS(content)
      outputPath = outputPath.replace('.ts', '.js')
    }

    return {
      [outputPath]: content,
      ...acc,
    }
  }, {})
}

export const builder = {
  crud: { type: 'boolean', default: false, desc: 'Create CRUD functions' },
  force: { type: 'boolean', default: false },
  typescript: {
    type: 'boolean',
    default: false,
    desc: 'Generate TypeScript files',
  },
  javascript: {
    type: 'boolean',
    default: true,
    desc: 'Generate JavaScript files',
  },
}

export const { command, desc, handler } = createYargsForComponentGeneration({
  componentName: 'service',
  filesFn: files,
  builder,
})
