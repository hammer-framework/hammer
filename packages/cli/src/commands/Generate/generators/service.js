import path from 'path'

import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'

import { generateTemplate } from 'src/lib'

const OUTPUT_PATH = path.join('api', 'src', 'services')

const files = (args) => {
  const [[serviceName, ..._rest], flags] = args
  const singularName = pascalcase(serviceName)
  const pluralName = pluralize(singularName)
  const singularCamelName = camelcase(singularName)
  const pluralCamelName = camelcase(pluralName)
  const outputPath = path.join(OUTPUT_PATH, `${pluralCamelName}.js`)
  const isCrud = !!flags['crud']
  const template = generateTemplate(
    path.join('service', 'service.js.template'),
    { singularName, pluralName, singularCamelName, pluralCamelName, isCrud }
  )

  return { [outputPath]: template }
}

export default {
  name: 'Service',
  command: 'service',
  description: 'Generates a service object',
  files: (args) => files(args),
}
