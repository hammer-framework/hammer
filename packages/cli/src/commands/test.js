import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import { getPaths } from 'src/lib'
import c from 'src/lib/colors'

export const command = 'test [app..]'
export const desc = 'Run Jest tests for api and web.'
export const builder = {
  app: { choices: ['api', 'web'], default: ['api', 'web'] },
}

export const handler = async ({ app }) => {
  const { base: BASE_DIR } = getPaths()
  const execCommands = {
    api: {
      cwd: `${BASE_DIR}/api`,
      cmd: 'yarn jest',
      args: [
        '--passWithNoTests',
        '--config ../node_modules/@redwoodjs/core/config/jest.config.api.js',
      ],
    },
    web: {
      cwd: `${BASE_DIR}/web`,
      cmd: 'yarn jest',
      args: [
        '--passWithNoTests',
        '--config ../node_modules/@redwoodjs/core/config/jest.config.web.js',
      ],
    },
  }

  const tasks = new Listr(
    app.map((appName) => {
      const { cmd, args, cwd } = execCommands[appName]
      return {
        title: `Running '${appName}' jest tests`,
        task: () => {
          return execa(cmd, args, {
            stdio: 'inherit',
            shell: true,
            cwd,
          })
        },
      }
    }),
    {
      renderer: VerboseRenderer,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}
