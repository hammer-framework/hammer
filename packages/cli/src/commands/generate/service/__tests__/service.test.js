global.__dirname = __dirname
import { loadGeneratorFixture } from 'src/lib/test'

import * as service from '../service'

test('returns exactly 2 files', async () => {
  const files = await service.files({
    name: 'User',
    crud: false,
    relations: [],
  })

  expect(Object.keys(files).length).toEqual(2)
})

test('creates a single word service file', async () => {
  const files = await service.files({
    name: 'User',
    crud: false,
    relations: [],
  })

  expect(files['/path/to/project/api/src/services/users/users.js']).toEqual(
    loadGeneratorFixture('service', 'singleWord.js')
  )
})

test('creates a single word service test file', async () => {
  const files = await service.files({
    name: 'User',
    crud: false,
    relations: null,
  })

  expect(
    files['/path/to/project/api/src/services/users/users.test.js']
  ).toEqual(loadGeneratorFixture('service', 'singleWord.test.js'))
})

test('creates a multi word service file', async () => {
  const files = await service.files({
    name: 'UserProfile',
    crud: false,
    relations: null,
  })

  expect(
    files['/path/to/project/api/src/services/userProfiles/userProfiles.js']
  ).toEqual(loadGeneratorFixture('service', 'multiWord.js'))
})

test('creates a multi word service test file', async () => {
  const files = await service.files({
    name: 'UserProfile',
    crud: false,
    relations: null,
  })

  expect(
    files['/path/to/project/api/src/services/userProfiles/userProfiles.test.js']
  ).toEqual(loadGeneratorFixture('service', 'multiWord.test.js'))
})

test('creates a single word service file with CRUD actions', async () => {
  const files = await service.files({
    name: 'Post',
    crud: true,
    relations: null,
  })

  expect(files['/path/to/project/api/src/services/posts/posts.js']).toEqual(
    loadGeneratorFixture('service', 'singleWord_crud.js')
  )
})

test('creates a service test file with CRUD actions', async () => {
  const files = await service.files({
    name: 'Post',
    crud: true,
    relations: null,
  })

  expect(
    files['/path/to/project/api/src/services/posts/posts.test.js']
  ).toEqual(loadGeneratorFixture('service', 'singleWord_crud.test.js'))
})

test('creates a multi word service file with CRUD actions', async () => {
  const files = await service.files({
    name: 'UserProfile',
    crud: true,
    relations: null,
  })

  expect(
    files['/path/to/project/api/src/services/userProfiles/userProfiles.js']
  ).toEqual(loadGeneratorFixture('service', 'multiWord_crud.js'))
})

test('creates a multi word service test file with CRUD actions', async () => {
  const files = await service.files({
    name: 'UserProfile',
    crud: true,
    relations: null,
  })

  expect(
    files['/path/to/project/api/src/services/userProfiles/userProfiles.test.js']
  ).toEqual(loadGeneratorFixture('service', 'multiWord_crud.test.js'))
})

test('creates a single word service file with a hasMany relation', async () => {
  const files = await service.files({
    name: 'User',
    crud: false,
    relations: ['userProfiles'],
  })

  expect(files['/path/to/project/api/src/services/users/users.js']).toEqual(
    loadGeneratorFixture('service', 'singleWord_hasMany.js')
  )
})

test('creates a single word service file with a belongsTo relation', async () => {
  const files = await service.files({
    name: 'User',
    crud: false,
    relations: ['identity'],
  })

  expect(files['/path/to/project/api/src/services/users/users.js']).toEqual(
    loadGeneratorFixture('service', 'singleWord_belongsTo.js')
  )
})

test('creates a single word service file with multiple relations', async () => {
  const files = await service.files({
    name: 'User',
    crud: false,
    relations: ['userProfiles', 'identity'],
  })

  expect(files['/path/to/project/api/src/services/users/users.js']).toEqual(
    loadGeneratorFixture('service', 'singleWord_multiple.js')
  )
})
