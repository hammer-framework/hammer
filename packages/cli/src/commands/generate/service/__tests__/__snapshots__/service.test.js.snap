// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in javascript mode creates a multi word service file 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const userProfiles = () => {
  return db.userProfile.findMany()
}
"
`;

exports[`in javascript mode creates a multi word service test file 1`] = `
"import { userProfiles } from './userProfiles'

describe('userProfiles', () => {
  scenario('returns all userProfiles', async (scenario) => {
    const result = await userProfiles()

    expect(result.length).toEqual(Object.keys(scenario.userProfile).length)
  })
})
"
`;

exports[`in javascript mode creates a single word service file 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}
"
`;

exports[`in javascript mode creates a single word service file with CRUD actions 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const posts = () => {
  return db.post.findMany()
}

export const post = ({ id }) => {
  return db.post.findUnique({
    where: { id },
  })
}

export const createPost = ({ input }) => {
  return db.post.create({
    data: input,
  })
}

export const updatePost = ({ id, input }) => {
  return db.post.update({
    data: input,
    where: { id },
  })
}

export const deletePost = ({ id }) => {
  return db.post.delete({
    where: { id },
  })
}
"
`;

exports[`in javascript mode creates a single word service file with a belongsTo relation 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  identity: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in javascript mode creates a single word service file with a hasMany relation 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
}
"
`;

exports[`in javascript mode creates a single word service file with multiple relations 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
  identity: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in javascript mode creates a single word service test file 1`] = `
"import { users } from './users'

describe('users', () => {
  scenario('returns all users', async (scenario) => {
    const result = await users()

    expect(result.length).toEqual(Object.keys(scenario.user).length)
  })
})
"
`;

exports[`in typescript mode creates a multi word service file 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'
import { BeforeResolverSpecType } from '@redwoodjs/api'

export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const userProfiles = () => {
  return db.userProfile.findMany()
}
"
`;

exports[`in typescript mode creates a multi word service test file 1`] = `
"import { userProfiles } from './userProfiles'

describe('userProfiles', () => {
  scenario('returns all userProfiles', async (scenario) => {
    const result = await userProfiles()

    expect(result.length).toEqual(Object.keys(scenario.userProfile).length)
  })
})
"
`;

exports[`in typescript mode creates a single word service file 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'
import { BeforeResolverSpecType } from '@redwoodjs/api'

export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}
"
`;

exports[`in typescript mode creates a single word service file with CRUD actions 1`] = `
"import { Prisma } from '@prisma/client'
import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'
import { BeforeResolverSpecType } from '@redwoodjs/api'

export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const posts = () => {
  return db.post.findMany()
}

export const post = ({ id }: Prisma.PostWhereUniqueInput) => {
  return db.post.findUnique({
    where: { id },
  })
}

interface CreatePostArgs {
  input: Prisma.PostCreateInput
}

export const createPost = ({ input }: CreatePostArgs) => {
  return db.post.create({
    data: input,
  })
}

interface UpdatePostArgs extends Prisma.PostWhereUniqueInput {
  input: Prisma.PostUpdateInput
}

export const updatePost = ({ id, input }: UpdatePostArgs) => {
  return db.post.update({
    data: input,
    where: { id },
  })
}

export const deletePost = ({ id }: Prisma.PostWhereUniqueInput) => {
  return db.post.delete({
    where: { id },
  })
}
"
`;

exports[`in typescript mode creates a single word service file with a belongsTo relation 1`] = `
"import { Prisma } from '@prisma/client'
import { ResolverArgs } from '@redwoodjs/api/dist/types'
import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'
import { BeforeResolverSpecType } from '@redwoodjs/api'

export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  identity: (_obj, { root }: ResolverArgs<Prisma.UserWhereUniqueInput>) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in typescript mode creates a single word service file with a hasMany relation 1`] = `
"import { Prisma } from '@prisma/client'
import { ResolverArgs } from '@redwoodjs/api/dist/types'
import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'
import { BeforeResolverSpecType } from '@redwoodjs/api'

export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }: ResolverArgs<Prisma.UserWhereUniqueInput>) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
}
"
`;

exports[`in typescript mode creates a single word service file with multiple relations 1`] = `
"import { Prisma } from '@prisma/client'
import { ResolverArgs } from '@redwoodjs/api/dist/types'
import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'
import { BeforeResolverSpecType } from '@redwoodjs/api'

export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }: ResolverArgs<Prisma.UserWhereUniqueInput>) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
  identity: (_obj, { root }: ResolverArgs<Prisma.UserWhereUniqueInput>) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in typescript mode creates a single word service test file 1`] = `
"import { users } from './users'

describe('users', () => {
  scenario('returns all users', async (scenario) => {
    const result = await users()

    expect(result.length).toEqual(Object.keys(scenario.user).length)
  })
})
"
`;
